### PR TITLE
HPCC-15660 Can't pass stored values to EMBED options in Thor

### PR DESCRIPTION
**Pull request by richardkchapman**
Rebased to 6.0.2

### CODE REVIEW

**Comment by richardkchapman** [comment]
@ghalliday Please review

**Comment by ghalliday** [comment]
streamread.ecl shows that the option is being added to c++ functions that use embeded c++.  I'm not sure if it is a big issue, but it isn't necessary.

**Comment by ghalliday** [comment]
The options contain system options (e.g., $_linkcounted_), which should probably be excluded.

**Comment by ghalliday** [comment]
The generated ecl also includes the new options - a minor issue, but slightly confusing.

(E.g., embedR.ecl.xml)

**Comment by ghalliday** [comment]
Similarly ecl/regress/issue12854.ecl:

         <att name="ecl" value="getallstatus();&#10;"/>

         <att name="ecl" value="getallstatus(INTERNAL(&apos;gl2&apos;));&#10;"/>

**Comment by ghalliday** [comment]
@richardkchapman we should remove the argument from c++ embeds.

**Comment by richardkchapman** [comment]
Removing the system options should be a separate Jira (it's not a new issue and I'd need to check that no plugins ever actually looked at them...). I have fixed the other two issues. If you re-review I can squash before you merge

**Comment by richardkchapman** [comment]
@ghalliday Back to you

File: ecl/hql/hqlthql.cpp [hunk 1]
```diff
@@ -453,16 +453,23 @@ StringBuffer &HqltHql::callEclFunction(StringBuffer &s, IHqlExpression * expr, b
...
     s.append('(');
     unsigned numParameters = formals->numChildren();
+    bool first = true;
     for (unsigned idx = 0; idx < numParameters; idx++)
     {
-        if (idx)
-            s.append(", ");
-        
-        if(funcdef->isNamedSymbol())
+        if (!formals->queryChild(idx)->hasAttribute(_hidden_Atom))
```
Comment (ghalliday) [inline ecl/hql/hqlthql.cpp]: please add 

   expandProcessed ||

to these tests, so that when the expression tree is traced they are present.

**Comment by ghalliday** [comment]
@richardkchapman quite a few examples now fail.  E.g., apply.ecl.

Please run the compiler regression suite (see ecl/eclcc/DOCUMENTATION.rst for details) on the two public regression suites before releasing!

File: ecl/hqlcpp/hqlttcpp.cpp [hunk 2]
```diff
@@ -6216,18 +6216,85 @@ IHqlExpression * WorkflowTransformer::transformInternalFunction(IHqlExpression *
...
+            // Check that they didn't and give a sensible error message
+            OwnedHqlExpr boundCast = replaceInlineParameters(newFuncDef, cast);
+            if (cast != boundCast)
+                throwError(HQLERR_EmbedParamNotSupportedInOptions);
+
+            folded.setown(foldHqlExpression(cast));
+        }
+        else
+            folded.setown(createConstant(""));
+        HqlExprArray newFormals;
```
Comment (ghalliday) [inline ecl/hqlcpp/hqlttcpp.cpp]: trivial style: I would add a newline to separate the logic, and possibly have the previous code in a separate function

File: ecl/hqlcpp/hqlttcpp.cpp [hunk 3]
```diff
@@ -6216,18 +6216,85 @@ IHqlExpression * WorkflowTransformer::transformInternalFunction(IHqlExpression *
...
+        HqlExprArray attrs;
```
Comment (ghalliday) [inline ecl/hqlcpp/hqlttcpp.cpp]: trivial style: I would move this down a line so it is closer to its use - it makes the code slightly easier to read

File: ecl/hqlcpp/hqlttcpp.cpp [hunk 4]
```diff
@@ -6216,18 +6216,85 @@ IHqlExpression * WorkflowTransformer::transformInternalFunction(IHqlExpression *
...
+        attrs.append(*createAttribute(_hidden_Atom));
+        newFormals.append(*createParameter(__optionsId, newFormals.length(), LINK(unknownVarStringType), attrs));
+
+        HqlExprArray newDefaults;
+        if (defaults)
+            unwindChildren(newDefaults, defaults, 0);
+        while (newDefaults.length() < formals->numChildren())
+            newDefaults.append(*createOmittedValue());
+        newDefaults.append(*folded.getClear());
+        formals = formals->clone(newFormals);
```
Comment (ghalliday) [inline ecl/hqlcpp/hqlttcpp.cpp]: I think both formals and defaults leak

Comment (ghalliday) [inline ecl/hqlcpp/hqlttcpp.cpp]: trivial style: I would move this up just after the newFormals append so code is more tightly together

File: ecl/hqlcpp/hqlttcpp.cpp [hunk 5]
```diff
@@ -6216,18 +6216,85 @@ IHqlExpression * WorkflowTransformer::transformInternalFunction(IHqlExpression *
...
+        defaults = createValueSafe(no_sortlist, makeSortListType(NULL), newDefaults);
+        funcdefArgs.append(*LINK(formals));
+        funcdefArgs.append(*LINK(defaults));
+        unwindChildren(funcdefArgs, newFuncDef, 3);
+        OwnedHqlExpr namedFuncDef = newFuncDef->clone(funcdefArgs);
```
Comment (ghalliday) [inline ecl/hqlcpp/hqlttcpp.cpp]: trivial:  The end of this code code be commoned up with the else branch.

**Comment by ghalliday** [comment]
@richardkchapman I think the leak needs fixing, the others comments are trivial.  Please can you squash afterwards.

File: ecl/hqlcpp/hqlttcpp.cpp (hunk 5 above)
Comment (richardkchapman) [inline ecl/hqlcpp/hqlttcpp.cpp]: But would it make it clearer/more easily maintained. I thought not...

**Comment by richardkchapman** [comment]
Fixed the leak and some of the style issues, and squashed

@ghalliday

**Pull request merged**